staging/src/k8s.io/apiserver/pkg/cel/library/image.go

```diff
@@ -0,0 +1,317 @@
+//
+//	image(<string>) <Image>
+//
+// Examples:
+//
+//	image('registry.k8s.io/kube-apiserver-arm64:latest') // Returns an Image
+//	image('nginx') // Returns an Image
+//	image('registry.k8s.io/kube-apiserver-arm64@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2') // Returns an Image
+//	image('registry.k8s.io/kube-apiserver-arm64:latest@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2') // Returns an Image
+//	image('registry.k8s.io//kube-apiserver-arm64') // error
+//	image(0) // error
+///	image('ubuntu') // Returns Image for "ubuntu".
+///	image('ubuntu', true) // Applies normalization and returns components of "docker.io/library/ubuntu".
+///	image('ubuntu', false) // Returns Image for "ubuntu".
+//	image('registry.k8s.io/kube-apiserver-arm64:latest', true) // Returns an Image
+//
+// isImage
+//
+// Returns true if a string is a valid image. isImage returns true if and
+// only if image reference parsing does not result in error.
+//
+//	isImage(<string>) <bool>
+//
+// Examples:
+//
+//	isImage('registry.k8s.io/kube-apiserver-arm64:latest') // Returns true
+//	isImage('nginx') // Returns true
+//	isImage('registry.k8s.io/kube-apiserver-arm64@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2') // Returns true
+//	isImage('registry.k8s.io/kube-apiserver-arm64:latest@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2') // Returns true
+//	isImage('registry.k8s.io//kube-apiserver-arm64') // Returns false
+//	isImage(0) // Returns false
+//
+// <Image>.containsDigest:
+//
+// Returns true if the Image has a digest.
+//
+// Examples:
+//
+//	image("image("registry.k8s.io/kube-apiserver-arm64:latest@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2").containsDigest()").containsDigest() // Returns true
+//	image("registry.k8s.io/kube-apiserver-arm64@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2").containsDigest()) // Returns true
+//	image("registry.k8s.io/kube-apiserver-arm64:latest").containsDigest() // Returns false
+//	image("registry.k8s.io/kube-apiserver-arm64").containsDigest() // Returns false
+//
+// Image components:
+//
+//   - registry: Returns the registry component of the image reference.
+//
+//   - repository: Returns the repository component of the image reference.
+//
+//   - tag: Returns the registry component of the image reference. Defaults to latest if both tag and digest are not present.
+//
+//   - digest: Returns the digest component of the image reference.
+//
+// Examples:
+//
+// image("kube-apiserver-arm64:testtag").registry() // Returns ""
+// image("registry.k8s.io/kube-apiserver-arm64:latest").registry() // Returns "registry.k8s.io"
+// image("registry.k8s.io/kube-apiserver-arm64").repository() // Returns "kube-apiserver-arm64"
+// image("registry.k8s.io/kube-apiserver-arm64").tag() // Returns "latest"
```

Comment:
It's interesting that this performs implicit defaulting (returning "latest" when the tag is unset).  I'm trying to get my head around the edge cases:

- Does this still return "latest" if there is also a digest set?  That could be misleading.
- Why does this perform implicit defaulting while `registry()` does not (`docker.io` is the default if the registry is not set, right?)

Reply:
yeah... we don't default a missing registry to docker.io in the API layer, we plumb the image verbatim all the way down to the node, which is where the `latest` tag is added by the kubelet, and the default registry is added by the container runtime

Reply:
TIL, I could've sworn we defaulted it to docker.io at least in kubelet before it got to CRI.

runtimes do sometimes rewrite it at the network layer too though either way

personally if I were strictly enforcing image policies about hosts etc I'd probably want that ensured in the container runtime. But a policy like "must be valid" or "must use a digest" seems reasonable to desire out at pod / daemonset / ...

Reply:
> TIL, I could've sworn we defaulted it to docker.io at least in kubelet before it got to CRI.

not intentionally... it regressed a couple times and got clarified that image name should be passed down to CRI as-is

https://github.com/kubernetes/kubernetes/issues/52110#issuecomment-361214391

https://github.com/kubernetes/kubernetes/pull/58955

Reply:
Any library we add should not bake in registry or tag defaulting that actually depends on the container runtime configuration. Making it visible that there is no registry or tag or digest on an image (with the possibility to return "") could be ok, so someone could write a policy that requires it be set, etc

Reply:
Good to know, there's also an implicit default of `ubuntu` => `docker.io/library/ubuntu:latest`, I haven't seen `/library` mentioned in this thread yet.

Reply:
I missed that case in the documentation. If image is not normalized (default behaviour), we pass the image as it is
```
		{
			name: "parse_non_normalized",
			expr: `image("kube-apiserver-arm64", false)`,
			expectValue: apiservercel.Image{
				ImageReference: apiservercel.ImageReference{
					Image:      "kube-apiserver-arm64",
					Registry:   "",
					Repository: "kube-apiserver-arm64",
					Tag:        "",
					Identifier: "",
				},
			},
		},
```